### PR TITLE
faster replacements for isFunctional,Function,TotalPattern

### PR DESCRIPTION
**Pull request by u--**
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: src/main/haskell/kore/src/Kore/Step/RecursiveAttributes.hs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+recursivelyCheckHeadProperty prop tools = go
+  where
+    go (App_ patHead patChildren) = prop atts && all go patChildren
+        where atts = MetadataTools.symAttributes tools patHead
+    go (DV_ _ pat) = all go pat
+    go (Var_ _)           = True
+    go (StringLiteral_ _) = True
+    go (CharLiteral_ _)   = True
+    go _ = False
+    -- go pat = all go $ pat ^. partsOf allChildren
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Step/RecursiveAttributes.hs]: This comment seems not to be needed:
```suggestion

```

Comment (ttuegel) [reply src/main/haskell/kore/src/Kore/Step/RecursiveAttributes.hs]: Also, this "Suggested change" feature is kinda cool.

**Comment by traiansf** [comment]
I understand.  My issue with it is that I changed something in `isFunctionalPattern`, but since the code was not removed, I did not get any conflicts but my tests started suddenly failing after rebasing.  Luckily I remembered that @u-- was working on reimplementing them...

File: src/main/haskell/kore/src/Kore/Step/RecursiveAttributes.hs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  where
+    go (App_ patHead patChildren) = prop atts && all go patChildren
+        where atts = MetadataTools.symAttributes tools patHead
+    go (DV_ _ pat) = all go pat
+    go (Var_ _)           = True
+    go (StringLiteral_ _) = True
+    go (CharLiteral_ _)   = True
+    go _ = False
+
+isFunctionalPattern, isFunctionPattern, isTotalPattern
```
Comment (virgil-serbanuta) [inline src/main/haskell/kore/src/Kore/Step/RecursiveAttributes.hs]: Missing documentation.

**Pull request merged**